### PR TITLE
Feat/stacks signer fee estimate

### PR DESCRIPTION
**Pull request by jferrant**
Closes https://github.com/stacks-network/stacks-core/issues/4543

### CODE REVIEW

File: stacks-signer/src/signer.rs [hunk 1]
```diff
@@ -517,31 +517,29 @@ impl Signer {
                 sig: vec![],
             };
             self.handle_packets(stacks_client, res, &[packet], current_reward_cycle);
+        } else if block_info.valid.unwrap_or(false)
```
Comment (jcnelson) [inline stacks-signer/src/signer.rs:520]: I'm curious why this code path is preserved?  I thought the Stacks node miner itself would always be the signing round coordinator for a block?

**Review by jcnelson — DISMISSED** [review]
Overall LGTM, just had a couple of questions that need to be answered prior to merging

**Review by obycode — DISMISSED** [review]

File: stacks-signer/src/signer.rs (hunk 1 above)
Comment (jferrant) [reply stacks-signer/src/signer.rs:520]: Good catch. This was just me fixing a clippy warning without looking into what the branch was. Will remove.

File: stacks-signer/src/signer.rs [hunk 2]
```diff
@@ -1054,6 +1032,44 @@ impl Signer {
...
+            self.reward_cycle,
+            nonce,
+        )?;
+        let tx_fee = if epoch < &StacksEpochId::Epoch30 {
+            debug!("{self}: in pre Epoch 3.0 cycles, must set a transaction fee for the DKG vote.");
+            let fee = if let Some(max_fee) = self.max_tx_fee_ustx {
+                let estimated_fee = stacks_client
+                    .get_medium_estimated_fee_ustx(&unsigned_tx)
+                    .map_err(|e| {
+                        debug!("{self}: unable to estimate fee for DKG vote transaction: {e:?}");
```
[the remark this answers is not in the document]
Comment (jferrant) [reply stacks-signer/src/signer.rs:1056]: Done :)

File: stacks-signer/src/signer.rs (hunk 1 above)
Comment (jferrant) [reply stacks-signer/src/signer.rs]: @kantai could you confirm that me removing this is okay? I assumed it was just overlooked, but want to make sure we are not anticipating that signers operate as acoordinator if for some reason they never set the self.miner key?

Comment (kantai) [reply stacks-signer/src/signer.rs]: Removing this should be okay, yes.

**Review by jcnelson — APPROVED** [review]

File: stacks-signer/src/signer.rs (hunk 1 above)
Comment (jferrant) [reply stacks-signer/src/signer.rs]: Just realized in this thought exercise...I am not properly handling DKG triggered in the middle of a current reward cycle still. A signer might try to trigger it as it should. But all other signers will ignore it as they treat the message as not coming from the miner coordinator.

**Review by hstove — CHANGES_REQUESTED** [review]
Everything looks good except the logging of the whole config at startup - I really think we shouldn't have that logged as-is.

File: stacks-signer/src/client/stacks_client.rs [hunk 3]
```diff
@@ -196,9 +198,43 @@ impl StacksClient {
...
+        };
+        let send_request = || {
+            self.stacks_node_client
+                .post(self.fees_transaction_path())
+                .header("Content-Type", "application/json")
+                .json(&request)
+                .send()
+                .map_err(backoff::Error::transient)
+        };
+        let response = retry_with_exponential_backoff(send_request)?;
```
[the remark this answers is not in the document]
Comment (jferrant) [reply stacks-signer/src/client/stacks_client.rs:218]: A 400 does not trigger a retry. Only errors return a retry (like reqwest errors)

File: stacks-signer/src/config.rs [hunk 4]
```diff
@@ -394,7 +411,8 @@ mod tests {
         let network = Network::Testnet;
         let password = "melon";
 
-        let config_tomls = build_signer_config_tomls(&[pk], node_host, None, &network, password);
+        let config_tomls =
+            build_signer_config_tomls(&[pk], node_host, None, &network, password, None);
```
[the remark this answers is not in the document]
Comment (jferrant) [reply stacks-signer/src/config.rs:415]: It will not be set. Its an optional.

**Review by hstove — DISMISSED** [review]
LGTM!

**Review by kantai — DISMISSED** [review]

**Review by jcnelson — DISMISSED** [review]

File: stacks-signer/src/signer.rs [hunk 5]
```diff
@@ -1071,6 +1025,44 @@ impl Signer {
...
+    ) -> Result<StacksTransaction, ClientError> {
+        let mut unsigned_tx = stacks_client.build_unsigned_vote_for_aggregate_public_key(
+            self.stackerdb.get_signer_slot_id().0,
+            self.coordinator.current_dkg_id,
+            dkg_public_key,
+            self.reward_cycle,
+            nonce,
+        )?;
+        let tx_fee = if epoch < &StacksEpochId::Epoch30 {
+            debug!("{self}: in pre Epoch 3.0 cycles, must set a transaction fee for the DKG vote.");
```
Comment (jcnelson) [inline stacks-signer/src/signer.rs:1044]: Might want to make this `INFO` so it's more discoverable by operators

**Review by jcnelson — DISMISSED** [review]

**Review by kantai — DISMISSED** [review]

**Review by kantai — APPROVED** [review]

**Review by jcnelson — APPROVED** [review]

**Comment by obycode** [comment]
Should we go ahead and merge this, or re-target it to develop?

**Pull request merged**